File: everything_parsing/25-ll1-table/java/LL1TableTest.java
import java.util.List;

/** java -cp java/out LL1TableTest */
public final class LL1TableTest {

    static LL1Table load(String g) { return new LL1Table(Grammar.parse(g)); }

    public static void main(String[] args) {
        System.out.println("LL1TableTest");

        LL1Table e = load("""
            %start E
            E  -> T Ep
            Ep -> + T Ep | epsilon
            T  -> F Tp
            Tp -> * F Tp | epsilon
            F  -> ( E ) | id | num
            """);

        // ── the table is single-valued ──
        Assert.that(e.isLL1(), "the expression grammar's table has no conflicts");
        Assert.equals(e.conflicts().size(), 0, "zero conflicted cells");

        // ── specific cells ──
        Assert.equals(e.cell("E", "id").get(0).index(), 0, "M[E][id] = production 0 (E -> T Ep)");
        Assert.equals(e.cell("E", "(").get(0).index(), 0, "M[E][(] = 0");
        Assert.equals(e.cell("Ep", "+").get(0).index(), 1, "M[Ep][+] = 1 (Ep -> + T Ep)");
        Assert.equals(e.cell("Ep", ")").get(0).index(), 2, "M[Ep][)] = 2 (Ep -> epsilon), from FOLLOW");
        Assert.equals(e.cell("Ep", "$").get(0).index(), 2, "M[Ep][$] = 2");
        Assert.that(e.cell("E", "+").isEmpty(), "M[E][+] is blank -> syntax error on '+' where E is expected");
        Assert.that(e.cell("Tp", "*").get(0).index() == 4, "M[Tp][*] = 4 (Tp -> * F Tp)");

        // ── columns are terminals + $ ──
        Assert.that(e.columns.contains("$"), "$ is a column");
        Assert.that(!e.columns.contains("E"), "nonterminals are not columns");

        // ── dangling else: one conflicted cell ──
        LL1Table de = load("""
            %start S
            S  -> other | if b then S Sp
            Sp -> else S | epsilon
            """);
        Assert.that(!de.isLL1(), "left-factored dangling else is still not LL(1)");
        Assert.equals(de.conflicts().size(), 1, "exactly one conflicted cell");
        LL1Table.Conflict c = de.conflicts().get(0);
        Assert.equals(c.nt(), "Sp", "conflict is in row Sp");
        Assert.equals(c.token(), "else", "conflict is in column else");
        Assert.equals(c.ps().size(), 2, "two productions compete: Sp -> else S and Sp -> epsilon");

        // ── nullable list rule ──
        LL1Table st = load("""
            %start P
            P -> L
            L -> S L | epsilon
            S -> id assign E semi
            E -> id | num
            """);
        Assert.that(st.isLL1(), "statement grammar is LL(1)");
        Assert.equals(st.cell("L", "$").get(0).index(), 2, "M[L][$] = L -> epsilon (from FOLLOW(L))");
        Assert.equals(st.cell("L", "id").get(0).index(), 1, "M[L][id] = L -> S L");

        Assert.summary();
    }
}

File: everything_parsing/25-ll1-table/js/ll1table.test.mjs
// node js/ll1table.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { LL1Table } from "./ll1table.mjs";

console.log("ll1table.test.mjs");

const load = g => new LL1Table(Grammar.parse(g));

const e = load(`
%start E
E  -> T Ep
Ep -> + T Ep | epsilon
T  -> F Tp
Tp -> * F Tp | epsilon
F  -> ( E ) | id | num
`);

that(e.isLL1(), "the expression grammar's table has no conflicts");
equals(e.conflicts().length, 0, "zero conflicted cells");

equals(e.cell("E", "id")[0].index, 0, "M[E][id] = production 0");
equals(e.cell("E", "(")[0].index, 0, "M[E][(] = 0");
equals(e.cell("Ep", "+")[0].index, 1, "M[Ep][+] = 1");
equals(e.cell("Ep", ")")[0].index, 2, "M[Ep][)] = 2 (from FOLLOW)");
equals(e.cell("Ep", "$")[0].index, 2, "M[Ep][$] = 2");
that(e.cell("E", "+").length === 0, "M[E][+] is blank");
equals(e.cell("Tp", "*")[0].index, 4, "M[Tp][*] = 4");

that(e.columns.includes("$"), "$ is a column");
that(!e.columns.includes("E"), "nonterminals are not columns");

const de = load(`
%start S
S  -> other | if b then S Sp
Sp -> else S | epsilon
`);
that(!de.isLL1(), "left-factored dangling else is still not LL(1)");
equals(de.conflicts().length, 1, "exactly one conflicted cell");
const c = de.conflicts()[0];
equals(c.nt, "Sp", "conflict is in row Sp");
equals(c.token, "else", "conflict is in column else");
equals(c.ps.length, 2, "two productions compete");

const st = load(`
%start P
P -> L
L -> S L | epsilon
S -> id assign E semi
E -> id | num
`);
that(st.isLL1(), "statement grammar is LL(1)");
equals(st.cell("L", "$")[0].index, 2, "M[L][$] = L -> epsilon");
equals(st.cell("L", "id")[0].index, 1, "M[L][id] = L -> S L");

summary();
